```{r echo=FALSE}
library(knitr)
opts_chunk$set(echo=FALSE, message = FALSE, warning = FALSE)
options(stringsAsFactors = FALSE)
source("functions-analysis.R")
```


---
title: "TE count QTL"
author: "Lu Wang"
output: html_document
---


```{r}
c5 = c("#3953A4", "#6ABD45", "#ED2124", "#FAA41A", "#D16A27")
t5 = col2alpha(c5, alpha=0.2)
names(c5) = c("African", "American", "Asian", "European", "Indian")
```


```{r Heterozygosity, resutls = 'hide', fig.height=3, fig.width=3}
# Sample population info
sample_info=read.table("./analysis_input/sampleInfo.txt", header=T)
sample_info_pop = split(sample_info, f=sample_info$group)
pop_individuals = lapply(sample_info_pop, function(d){ d$id })
n_pop_individuals = unlist(lapply(pop_individuals, length))


# Total SNP count
snpCounts=read.table("./abilbeast_output/chunk_snp_count.txt", header=F)
colnames(snpCounts)=c("chunkName", names(pop_individuals))
snpCounts=snpCounts[, -1]
snpCounts=colSums(snpCounts)


# TE count
geneExp = read.table("./abilbeast_input/presenceExp.txt", header = T, stringsAsFactors = F)
pop_geneExp = lapply(pop_individuals, function(x) geneExp[, c("id", x)] )

########################
# Heterozygous SNP count

mapply(function(pop_name, geneExp, individuals, snpCount, col){
    heteroCount=read.table(paste0("./abilbeast_output/snps_heterozygous_", pop_name, ".txt"), header=F)
    heteroCount=heteroCount[, -1]
    colnames(heteroCount)=individuals
    heteroz=colSums(heteroCount)
    heteroz=heteroz/snpCount
    TE_classes=geneExp[, 1]
    TE_classes=gsub("all", "All classes", TE_classes)
    geneExp=geneExp[, -1]
    
    par(cex=0.7)
    for(i in 1:nrow(geneExp)){
        x=unlist(geneExp[i, ])
        y=heteroz
        plot(x, y, xlab="TE counts", ylab="Heterozygosity", col=col, main=paste(pop_name, TE_classes[i]), pch=16)
        m=lm(y~x)
        abline(m, col=col2inv(col))
        eq = bquote(atop(italic(y) == .(round(coef(m)[1], 3)) + ~"("*.(round(coef(m)[2], 3))*")" %.% italic(x), italic(r)^2 == .(round(summary(m)$r.squared, 4))))
        mtext(eq, side = 3, line = -3, cex = 0.7)
    }
}, names(pop_geneExp), pop_geneExp, pop_individuals, snpCounts, t5)

```


```{r}
########
# Variance explained
########

# Phenotypical distance

pheno_dist=lapply(pop_geneExp, function(d){
    all=unlist(d[1, -1]) # All TE insertions
    Alu=unlist(d[2, -1]) # Alu insertions
    mall=dist(all)
})


# Genetic relationship



```


```{r results='hide', fig.width=4, fig.height=4}
# AFR

myQQplot("./analysis_input/chrAllPvalues_AFR_pval.txt", "./analysis_input/chrAllPvalues_AFR_Permuted_pval.txt", n_probs=100000, p_cut=0.01, col=c5[1], title=names(c5)[1])


# EAS

myQQplot("./analysis_input/chrAllPvalues_EAS_pval.txt", "./analysis_input/chrAllPvalues_EAS_Permuted_pval.txt", n_probs=100000, p_cut=0.01, col=c5[3], title=paste(names(c5)[3]))

# EUR

myQQplot("./analysis_input/chrAllPvalues_EUR_pval.txt", "./analysis_input/chrAllPvalues_EUR_Permuted_pval.txt", n_probs=100000, p_cut=0.01, col=c5[4], title=names(c5)[4])

```



```{r ManhattanPlot, fig.height=3, fig.width = 8}

#pdf("manhattan_plots.pdf", height = 3, width = 5)

cutoff=8

# AFR

pvals = loadPvals("./analysis_input/chrAll_AFR_20150801_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[1])),title = paste(names(c5)[1], "observed"), by="none")
par(cex=0.7)
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[1])


pvals = loadPvals("./analysis_input/chrAll_AFR_Permuted_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[1])),title = paste(names(c5)[1], "permuted"), by="none")
#hist(-log10(pvals$pval), breaks=1000, border=NA, col=c5[1], xlab="-log10 P value", main="")
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[1])



# EAS

pvals = loadPvals("./analysis_input/chrAll_EAS_20150801_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[3])),title = paste(names(c5)[3], "observed"))
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[3])


pvals = loadPvals("./analysis_input/chrAll_EAS_Permuted_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[3])),title = paste(names(c5)[3], "permuted"))
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[3])

# EUR

pvals = loadPvals("./analysis_input/chrAll_EUR_20150801_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[4])),title = paste(names(c5)[3], "observed"))
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[4])

pvals = loadPvals("./analysis_input/chrAll_EUR_Permuted_pval.txt")
#sum(-log10(pvals$pval) > cutoff)
plotManhattan(pvals, col = c("darkgrey", paste(c5[4])),title = paste(names(c5)[3], "permuted"))
histLogFreq(-log10(pvals$pval), breaks=1000, col=c5[4])


# pvals_ucsc = data.frame(pvals$chr, pvals$pos, pvals$pos, -log10(pvals$pval))
# write.table(pvals_ucsc, "./analysis_output/AFR_pval.bedGraph", sep = "\t", quote = F, row.names = F, col.names = F)

# pvals_ucsc = data.frame(pvals$chr, pvals$pos, pvals$pos, -log10(pvals$pval))
# write.table(pvals_ucsc, "./analysis_output/EAS_pval.bedGraph", sep = "\t", quote = F, row.names = F, col.names = F)

# pvals_ucsc = data.frame(pvals$chr, pvals$pos, pvals$pos, -log10(pvals$pval))
# write.table(pvals_ucsc, "./analysis_output/EUR_pval.bedGraph", sep = "\t", quote = F, row.names = F, col.names = F)
#dev.off()

```



## Regress to recombination rate

Recombination hotspots are genomic regions with recombination rate > 10

```{r regressRecomb, fig.width=6, fig.height=6}
#recomb = read.table("./analysis_input/RecombRate.20150713.txt", header = T, comment.char = "")
recomb = loadDecode("./analysis_input/decodeSexAveraged/")
colnames(recomb) = c("chr", "chromStart", "chromEnd", "decodeAvg")

# AFR
pvals = loadPvals("./analysis_input/chrAll_AFR_20150801_pval.txt")
pvals = pvalJoinRecomb(pvals, recomb)
plotRecomb(pvals, title = "Distribution of decodeAvg", binwidth=1)
# Remove zero decodeAvg
plotRecomb(pvals[which(pvals$decodeAvg > 0),], title="log transformed decodeAvg after removing 0", binwidth=0.5, yscale = "log2")
plotRegress(pvals, title="AFR", col=t5[1])
ggplotRegress(pvals, title="AFR", col = t5[1], bin.data="decodeAvg")
plotRegress(pvals[which(pvals$decodeAvg >= 10), ], title="AFR hot spot", col=t5[1])

# EAS
pvals = loadPvals("./analysis_input/chrAll_EAS_20150801_pval.txt")
pvals = pvalJoinRecomb(pvals, recomb)
plotRegress(pvals, title="EAS", col=t5[3])
ggplotRegress(pvals, title="EAS", col = t5[3], bin.data="decodeAvg")
plotRegress(pvals[which(pvals$decodeAvg >= 10), ], title="EAS hot spot", col=t5[3])


# EUR
pvals = loadPvals("./analysis_input/chrAll_EUR_20150801_pval.txt")
pvals = pvalJoinRecomb(pvals, recomb)
plotRegress(pvals, title="EUR", col=t5[4])
ggplotRegress(pvals, title="EUR", col = t5[4], bin.data="decodeAvg")
plotRegress(pvals[which(pvals$decodeAvg >= 10), ], title="EUR hot spot", col=t5[4])

```


## Regress to mappability(sequence uniqueness)


```{r regressMappability, fig.width=6, fig.height=7, results='hide'}
# ##########
# # AFR
# pvals = loadPvals("./analysis_input/chrAll_AFR_pval.txt")
# res = summaryMappability(pvals[which(pvals$chr %in% c("chr1")), ], "./analysis_input/wgEncodeDukeMapabilityUniqueness20bp/")
# 
# #plotUniqueness(pvals, col=c5[1], title="AFR chr19-22")
# mapply(function(data, col, title){
#     plotUniqueness(data, col, title)
# }, res, col=rep(c5[1], length(res)), title=paste("AFR", names(res)))
# 
# 
# # EAS
# pvals = loadPvals("./analysis_input/chrAll_EAS_pval.txt")
# res = summaryMappability(pvals[which(pvals$chr %in% c("chr20", "chr21", "chr22")), ], "./analysis_input/wgEncodeDukeMapabilityUniqueness20bp/")
# 
# mapply(function(data, col, title){
#     plotUniqueness(data, col, title)
# }, res, col=rep(c5[3], length(res)), title=paste("EAS", names(res)))
# 
# 
# 
# # EUR
# pvals = loadPvals("./analysis_input/chrAll_EUR_pval.txt")
# res = summaryMappability(pvals[which(pvals$chr %in% c("chr20")), ], "./analysis_input/wgEncodeDukeMapabilityUniqueness20bp/")
# 
# mapply(function(data, col, title){
#     plotUniqueness(data, col, title)
# }, res, col=rep(c5[4], length(res)), title=paste("EUR", names(res)))

```


## Check SNPs near centromeres or telomere

```{r fig.width=4}
######
# Get centromere and telomere locations

gaps = read.table("./analysis_input/hg19_gap.txt", header = T, comment.char = "")
gaps = gaps[which(gaps$type %in% c("centromere", "telomere")), c(2:4,8)]
gaps = gaps[order(gaps$chrom, gaps$chromStart), ]
# Add flanking region of centromere and telomere to search for SNPs
spacer = 1e7
# 3' telomeres or centromeres
gaps$chromStart=ifelse(gaps$chromStart==0 & gaps$type=="telomere", gaps$chromStart, gaps$chromStart-spacer)
# 5' telomeres or dentromeres
gaps$chromEnd = ifelse(gaps$chromStart>0 & gaps$type=="telomere", gaps$chromEnd, gaps$chromEnd+spacer)

######
# AFR

pvals = loadPvals("./analysis_input/chrAll_AFR_20150801_pval.txt")
pvals = pvals[order(pvals$chr, pvals$pos), ]
pvals = pvalJoinGaps(pvals, gaps)

#pvals = loadPvals("./analysis_input/chrAll_AFR_pval.txt")

pvals = pvals[order(pvals$pval), ]
top_n=20
pvals=pvals[1:top_n, ]
pvals = addBins(pvals, by="pos", bin_size=1e6)

population="AFR"

####
# Expression data
exprs=read.table("./analysis_input/presenceExp.txt", header = T)
rownames(exprs)=exprs$id
exprs = exprs[,-1]

####
# Covariates data
covar=read.table("abilbeast_input/sampleInfo.txt", header = T)

####
# SNP genotype data
genotype_files=paste0("abilbeast_input/", gsub("chr", "", pvals$chr), "_", (pvals$bin_start+1), "_", pvals$bin_end, ".genotype")

# write.table(paste0("lwang336@abilbeast.biology.gatech.edu:~/workdir/1kg/TEcount_analysis/", genotype_files[!duplicated(genotype_files)], "\n./"), "temp/genotype_files.txt", quote=F, row.names=F, col.names=F)

pvals=split(pvals, f=1:nrow(pvals))
res=mapply(function(pval, genotype){
    require(dplyr)
    require(ggplot2)
    
    te=unlist(pval["te"])
    snp=unlist(pval["snp"])
    pvalue=unlist(pval["pval"])
    chr=unlist(pval["chr"])
    pos=unlist(pval["pos"])
    info=ifelse(is.null(pval$type), "", paste("\nnear", pval$type))
    
    ex=unlist(exprs[paste(te), ])
    df=data.frame(id=names(ex), expression=ex)
    
    genotype_data=read.delim(genotype, header = T, sep = "\t")
    genotype_data=genotype_data[which(genotype_data[, 1]==snp), ]
    genotype_data=unlist(genotype_data[, -1])
    df$genotype=genotype_data
    
    mdf=left_join(df, covar, by="id")
    mdf=mdf[which(mdf$group==population), ]
    
    g=ggplot(mdf, aes(x=factor(genotype), y=expression)) + 
      geom_point(size = 2, alpha = 1/2, position = position_jitter(width = 0.2), aes(col = factor(genotype))) +
      geom_boxplot(alpha = 1/2, outlier.size = 0) + 
      labs(title = paste(chr, ":", pos, info, "\n\n", snp, "vs", te, "in", population)) + 
      xlab(paste("SNP genotype\n\nP.value = ", format(pvalue, digits=4))) + 
      ylab(paste("copies of", te, "insertions")) +
      theme_bw() + 
      theme(panel.grid.major=element_blank(), 
            panel.grid.minor=element_blank(),
            plot.title=element_text(size=rel(0.9)))
    return(g)
#     return(mdf)
}, pvals, genotype_files, SIMPLIFY=F)

res=lapply(res, print)

```


```{r NCBI, eval = F, fig.height=3, fig.width = 8}
# require(dplyr)

# ncbi_eqtl = read.delim("./analysis_input/eqtl-150630-2258-10451.tab", sep = "\t", header = T )
# ncbi_eqtl = ncbi_eqtl[, -1]
# names(ncbi_eqtl)[2] = "snp_id"
# 
# #African
# pvals = loadPvals("./analysis_input/chrAll_AFR_pval.txt")
# pvals = left_join(pvals, ncbi_eqtl, by="snp")
# pvals = pvals[complete.cases(pvals),]
# pvals = pvals[order(pvals$te_snp_pval, pvals$P.value), ]

#write.csv(pvals, "./analysis_output/ncbi_AFR_eQTL_gene.csv", quote = F, row.names = F)
```



```{r GTEx, fig.height=6, fig.width = 12, eval = F}
# require(dplyr)
# require(RColorBrewer)
# gtex_eqtl = list()
# 
# gtex_eqtl[["Adipose_Subcutaneous"]] = addGTEx("./analysis_input/GTEx/Adipose_Subcutaneous.portal.eqtl")
# gtex_eqtl[["Artery_Aorta"]] = addGTEx("./analysis_input/GTEx/Artery_Aorta.portal.eqtl")
# gtex_eqtl[["Artery_Tibial"]] = addGTEx("./analysis_input/GTEx/Artery_Tibial.portal.eqtl")
# gtex_eqtl[["Esophagus_Mucosa"]] = addGTEx("./analysis_input/GTEx/Esophagus_Mucosa.portal.eqtl")
# gtex_eqtl[["Esophagus_Muscularis"]] = addGTEx("./analysis_input/GTEx/Esophagus_Muscularis.portal.eqtl")
# gtex_eqtl[["Heart_Left_Ventricle"]] = addGTEx("./analysis_input/GTEx/Heart_Left_Ventricle.portal.eqtl")
# gtex_eqtl[["Lung"]] = addGTEx("./analysis_input/GTEx/Lung.portal.eqtl")
# gtex_eqtl[["Muscle_Skeletal"]] = addGTEx("./analysis_input/GTEx/Muscle_Skeletal.portal.eqtl")
# gtex_eqtl[["Nerve_Tibial"]] = addGTEx("./analysis_input/GTEx/Nerve_Tibial.portal.eqtl")
# gtex_eqtl[["Skin"]] = addGTEx("./analysis_input/GTEx/Skin_Sun_Exposed_Lower_leg.portal.eqtl")
# gtex_eqtl[["Stomach"]] = addGTEx("./analysis_input/GTEx/Stomach.portal.eqtl")
# gtex_eqtl[["Thyroid"]] = addGTEx("./analysis_input/GTEx/Thyroid.portal.eqtl")
# gtex_eqtl[["Whole_Blood"]] = addGTEx("./analysis_input/GTEx/Whole_Blood.portal.eqtl")
# 
# ct = brewer.pal(11, "Spectral")
# pal <- colorRampPalette(ct)
# ct = pal(13)
# names(ct) = names(gtex_eqtl)
# 
# #African
# pvals = loadPvals("./analysis_input/chrAll_AFR_pval.txt")
# sel_gtex = lapply(gtex_eqtl, function(eqtls, list_name, color){
#     res = left_join(pvals, eqtls, by="snp")
#     res = res[complete.cases(res),]
#     res = res[order(res$pval, res$P_Val), ]
#     })
# for(i in 1:length(sel_gtex)){
#     highlight = unique(sel_gtex[[i]][,"snp"])
#     g = plotManhattan(pvals, col = c("darkgrey", paste(c5[1]), "red"),title = paste(names(c5)[1], names(sel_gtex)[i]), highlight = highlight)
#     print(g)
# }
# 
# 
# #Asian
# pvals = loadPvals("./analysis_input/chrall_EAS_pval.txt")
# sel_gtex = lapply(gtex_eqtl, function(eqtls, list_name, color){
#     res = left_join(pvals, eqtls, by="snp")
#     res = res[complete.cases(res),]
#     res = res[order(res$pval, res$P_Val), ]
#     })
# for(i in 1:length(sel_gtex)){
#     highlight = unique(sel_gtex[[i]][,"snp"])
#     g = plotManhattan(pvals, col = c("darkgrey", paste(c5[3]), "green"),title = paste(names(c5)[3], names(sel_gtex)[i]), highlight = highlight)
#     print(g)
# }
# 
# 
# #European
# pvals = loadPvals("./analysis_input/chrall_EUR_pval.txt")
# sel_gtex = lapply(gtex_eqtl, function(eqtls, list_name, color){
#     res = left_join(pvals, eqtls, by="snp")
#     res = res[complete.cases(res),]
#     res = res[order(res$pval, res$P_Val), ]
#     })
# for(i in 1:length(sel_gtex)){
#     highlight = unique(sel_gtex[[i]][,"snp"])
#     g = plotManhattan(pvals, col = c("darkgrey", paste(c5[4]), "steelblue"),title = paste(names(c5)[3], names(sel_gtex)[i]), highlight = highlight)
#     print(g)
# }
```


## Tracks

http://jordan.biology.gatech.edu/lulu/tracks/AFR_pval.bedGraph

http://jordan.biology.gatech.edu/lulu/tracks/EAS_pval.bedGraph

http://jordan.biology.gatech.edu/lulu/tracks/EUR_pval.bedGraph