```{r echo=FALSE}
library(knitr)
opts_chunk$set(echo=FALSE, message = FALSE, warning = FALSE)
```

---
title: "TE eQTL Paper"
author: "Lu Wang"
date: "Friday, December 19, 2014"
output: html_document
---


```{r}
source("function-paper.R")
options(stringsAsFactors = FALSE)
set.seed(12345)
c3 = c("salmon", "steelblue", "grey")
alpha = 0.5
c3.t = c(rgb(0.98, 0.5, 0.44, alpha), rgb(0.27, 0.5, 0.71, alpha), rgb(0.75, 0.75, 0.75, alpha))
```

```{r cache = TRUE}
SNP_file_name = "./input/snps.txt"
snps_location_file_name = "./input/snploc.txt"
expression_file_name = "./input/geneExp.txt"
gene_location_file_name = "./input/geneloc.txt"
covariates_file_name = "./input/covariates.txt"

snps <- read.table(SNP_file_name, header = T, stringsAsFactors = F)
snploc <- read.table(snps_location_file_name, header = T, stringsAsFactors = F)
geneExp <- read.table(expression_file_name, header = T, stringsAsFactors = F)
geneloc <- read.table(gene_location_file_name, header = T, stringsAsFactors = F)
covar <- read.table(covariates_file_name, header = T, stringsAsFactors = F)

#normGeneExp <- read.table(normalized_expression_file_name, header = T, stringsAsFactors = F)
```

### _Breakdown by Population_

```{r}
# Covariates
pop.labs = c("GBR", "FIN", "CEU", "YRI", "TSI")
group.labs = c("EUR", "YRI")
rownames(covar) = covar$id
values = grep("id", colnames(covar), invert = T)
df <- as.data.frame(t(covar[, values]))
par(cex = 0.7)
counts = table(factor(df$population, levels = 1:5, labels = pop.labs))
bp <- barplot(counts, xlab = "Population", ylab = "Number of Individuals")
text(y= counts+5, x= bp, labels=as.numeric(counts), xpd=TRUE)
df$group = as.factor(df$population)
levels(df$group) = c("EUR", "EUR", "EUR", "YRI", "EUR")
df = data.frame(sample = rownames(df), df)
covar2 = df
```


### Merge European Populations

```{r fig.width = 3}
par(cex = 0.7)
counts = table(df$group)
bp <- barplot(counts, xlab = "Population", ylab = "Number of Individuals", col = c("steelblue", "salmon"))
text(y= counts+10, x= bp, labels=as.numeric(counts), xpd=TRUE)
```


```{r}
# Chi-square test for skewness of SNP presence in EUR vs YRI
chi.test.p <- apply(valCols(snps), 1, myChisq, expected = covar2)
chi.test.df <- data.frame(snps = names(chi.test.p), chisq.p = chi.test.p)
chi.test.q <- p.adjust(chi.test.p, method = "BH")
chi.test.df$chisq.q = chi.test.q
#myChisq(valCols(snps)[1, ], expected = covar2)
```


```{r breakdownTwo, results='hide'}
sdf <- split(df, f = df$group)
names(sdf) = group.labs

pop.names = lapply(sdf, rownames)
pop.snps = lapply(pop.names, function(x) { df = snps[, c("id", x)] } )
```

```{r results='asis', cache = T}
require(knitr)

res <- sapply(pop.snps, summaryPoly)
res1 <- summaryPoly(snps)


res = cbind(res, Union = res1)
res = rbind(res, All = colSums(res))
res = formatTable1(res)
kable(res, format = "pandoc", caption = "Polymorphic Transposible Element Counts", padding = 2)
```

## Figure 1A. Boxplot of TE Allele Frequency


```{r filter0af}
require(reshape)
require(ggplot2)

af.cut = 0
pop.snps.filtered.0 = lapply(pop.snps, filterAF, af.cut)
snps.filtered.0 = filterAF(snps, af.cut)


res = lapply(pop.snps.filtered.0, getAF)
res = lapply(res, getType)
mdf = melt(res, id.vars = c("id", "af", "type"))
colnames(mdf) = gsub("L1", "group", colnames(mdf))

res1 = getAF(snps.filtered.0)
res1 = getType(res1)
res1$group = "Union"

df = rbind(mdf, res1)
df$group = factor(df$group, levels = c("EUR", "YRI", "Union"))
df$type = factor(df$type, levels = c("ALU", "L1", "SVA", "All"))

#ggplot(df, aes(x = type, y = af )) + geom_boxplot(aes(fill = group)) + theme_bw() + theme(panel.grid = element_blank()) + scale_fill_manual(values = c3 ) + labs(x = "Type of TE", y = "Allele Frequency") + geom_hline(yintercept = 0.05, linetype = "dashed", col = "red")                                                                                                             

#pdf("Figure1A.pdf")
ggplot(df, aes(x = af)) + geom_density(aes(col = group)) + theme_bw() + theme(panel.grid = element_blank()) + scale_color_manual(values = c3 ) + labs(x = "Allele Frequency", y = "Density") + geom_vline(xintercept = 0.05, linetype = "dashed", col = "red") + facet_wrap( ~ type, nrow = 2) 
#dev.off()

#pdf("Figure1A-intc0.25.pdf")
ggplot(df, aes(x = af)) + geom_density(aes(col = group)) + theme_bw() + theme(panel.grid = element_blank()) + scale_color_manual(values = c3 ) + labs(x = "Allele Frequency", y = "Density") + geom_vline(xintercept = 0.05, linetype = "dashed", col = "red") + facet_wrap( ~ type, nrow = 2) + coord_cartesian(xlim = c(0, 0.25))
#dev.off()

#ggplot(df, aes(x = af)) + stat_ecdf(aes(col = group)) + theme_bw() + theme(panel.grid = element_blank()) + scale_color_manual(values = c3 ) + labs(x = "Allele Frequency", y = "Cumulative Density") + geom_vline(xintercept = 0.05, linetype = "dashed", col = "red")  + facet_wrap( ~ type, nrow = 2)        
```


```{r cache = T}
library(VennDiagram)
sdf = split(mdf, f = mdf$type)
ssdf = lapply(sdf, function(x) {y = split(x, f = x$group); y = lapply(y, function(z) as.character(z$id))} )

#sapply(ssdf, function(x){ fname = tempfile(tmpdir = ".", fileext = ".tiff"); venn.diagram(x, fname , imagetype = "tiff", fill = c3[1:2], height = 1500, width = 1500 )})
       
paths = sapply(ssdf, function(x){ fname = tempfile(tmpdir = ".", fileext = ".png")
                          venn.diagram(x, fname , imagetype = "png", fill = c3[1:2], height = 1500, width = 1500 )
                          fname
                          })
paths = gsub('\\.\\\\', '', paths, perl = T)
```

`r names(ssdf)[1]`
<img src ="`r paths[1]`" height=300 width=300 />

`r names(ssdf)[2]`
<img src ="`r paths[2]`" height=300 width=300 />

`r names(ssdf)[3]`
<img src ="`r paths[3]`" height=300 width=300 />

`r names(ssdf)[4]`
<img src ="`r paths[4]`" height=300 width=300 />



## Figure 1B. Number of TE used in Analysis(AF>0.05)

```{r figure1B, fig.width = 4}
af.cut = 0.05
pop.snps.filtered = lapply(pop.snps, filterAF, af.cut)
snps.filtered = filterAF(snps, af.cut)


res = lapply(pop.snps.filtered, getAF)
res = lapply(res, getType)
mdf = melt(res, id.vars = c("id", "af", "type"))
colnames(mdf) = gsub("L1", "group", colnames(mdf))

res1 = getAF(snps.filtered)
res1 = getType(res1)
res1$group = "Union"

df = rbind(mdf, res1)
df$group = factor(df$group, levels = c("EUR", "YRI", "Union"))
df$type = factor(df$type, levels = c("ALU", "L1", "SVA", "All"))

res = table(df$type, df$group)
kable(res, format = "pandoc", caption = "Polymorphic Transposible Average Allele Frequency", padding = 2)

#pdf("Figure1B.pdf")
par(cex = 0.7)
counts = res
bp <- barplot(t(counts), xlab = "Type of Polymorphic TE", ylab = "Number of Polymorphic TE", col = c3 , beside = T)
text(y= t(counts)+200, x= bp, labels=t(counts), xpd=TRUE, srt = 90)
legend("top", legend = colnames(counts), fill = c3 )
#dev.off()
```


### eQTL Analysis

```{r eQTLfile, results='hide'}
sdf = split(covar2, covar2$group)
res <- lapply(sdf, function(x){ df = t(x)
                                df = data.frame(id = rownames(df), df)
                                if(x$population[1] == 4){
                                  df["gender", ]
                                }else{
                                  df[c("gender", "population"), ]
                                }})

mapply(write.table, res, file = paste0("./temp-paper/covariates", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )

# snps
pop.snps.filtered$all = snps.filtered
mapply(write.table, pop.snps.filtered, file = paste0("./temp-paper/snps", 1:3, ".txt"), quote = rep(FALSE, 3), row.names = rep(FALSE, 3), sep = rep("\t", 3) )

# snploc
pop.snp.names = lapply(pop.snps.filtered, function(x) x$id )
pop.snploc = lapply(pop.snp.names, function(x){ df = snploc
                                            rownames(df) = snploc$snp
                                            df[x, ]})
mapply(write.table, pop.snploc, file = paste0("./temp-paper/snploc", 1:3, ".txt"), quote = rep(FALSE, 3), row.names = rep(FALSE, 3), sep = rep("\t", 3) )

# Expression
pop.names = lapply(sdf, rownames)
pop.geneExp = lapply(pop.names, function(x) geneExp[, c("id", x)] )
mapply(write.table, pop.geneExp, file = paste0("./temp-paper/geneExp", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )
```

```{r runMatrixEQTL, results ='hide', cache = F}
# Global par

p.cut = 1e-2
cisDist = 1e6

# all Population

SNP_file_name = "./temp-paper/snps3.txt"
snps_location_file_name = "./temp-paper/snploc3.txt"


meq.all <- runME( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name = "./3.all_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold = p.cut, noFDRsaveMemory = F)

# write.csv(meq.all$all$eqlts, "meq.all.csv", quote = F, row.names = F)
# rm(me_ct.all)



me_ct.all <- runCisTrans( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name_cis = "3.all_cis_out.txt", output_file_name_tra = "3.all_tra_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold_cis = p.cut, pvOutputThreshold_tra = p.cut, cisDist = cisDist, noFDRsaveMemory = F)

# saveRDS(me_ct.all, "me_ct.all.Rds")
# par(cex = 0.7)
# pdf("me_ct.all.pdf")
# plot(me_ct.all)
# dev.off()
# rm(me_ct.all)

```

```{r EURandYRI}
# EUR


SNP_file_name = "./temp-paper/snps1.txt"
snps_location_file_name = "./temp-paper/snploc1.txt"
expression_file_name = "./temp-paper/geneExp1.txt"
covariates_file_name = "./temp-paper/covariates1.txt"

meq.eur <- runME( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name = "1.all_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold = p.cut, noFDRsaveMemory = F)


# saveRDS(meq.eur, "meq.eur.Rds")
# par(cex = 0.7)
# pdf("meq.eur.pdf")
# plot(meq.eur)
# dev.off()
# rm(meq.eur)


me_ct.eur <- runCisTrans( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name_cis = "1.all_cis_out.txt", output_file_name_tra = "1.all_tra_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold_cis = p.cut, pvOutputThreshold_tra = p.cut, cisDist = cisDist, noFDRsaveMemory = F)

# saveRDS(me_ct.eur, "me_ct.eur.Rds")
# par(cex = 0.7)
# pdf("me_ct.eur.pdf")
# plot(me_ct.eur)
# dev.off()
# rm(me_ct.eur)




# YRI 

SNP_file_name = "./temp-paper/snps2.txt"
snps_location_file_name = "./temp-paper/snploc2.txt"
expression_file_name = "./temp-paper/geneExp2.txt"
covariates_file_name = "./temp-paper/covariates2.txt"



meq.yri <- runME( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name = "2.all_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold = p.cut, noFDRsaveMemory = F)

# saveRDS(meq.yri, "meq.yri.Rds")
# par(cex = 0.7)
# pdf("meq.yri.pdf")
# plot(meq.yri)
# dev.off()
# rm(meq.yri)


me_ct.yri <- runCisTrans( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, output_file_name_cis = "2.all_cis_out.txt", output_file_name_tra = "2.all_tra_out.txt", useModel = modelLINEAR, par.plot = 'qqplot', verbose = F, pvOutputThreshold_cis = p.cut, pvOutputThreshold_tra = p.cut, cisDist = cisDist, noFDRsaveMemory = F)

# saveRDS(me_ct.yri, "me_ct.yri.Rds")x
# par(cex = 0.7)
# pdf("me_ct.yri.pdf")
# plot(me_ct.yri)
# dev.off()

```


```{r eval = F}
#YRI
write.csv(me_ct.yri$trans$eqtls, "me_ct.yri.tra.csv", quote = F, row.names = F)
write.csv(me_ct.yri$cis$eqtls, "me_ct.yri.cis.csv", quote = F, row.names = F)
rm(me_ct.yri)

meq.yri = readRDS("meq.yri.Rds")
write.csv(meq.yri$all$eqtls, "meq.yri.csv", quote = F, row.names = F)
rm(meq.yri)


#EUR
me_ct.eur = readRDS("me_ct.eur.Rds")
write.csv(me_ct.eur$trans$eqtls, "me_ct.eur.tra.csv", quote = F, row.names = F)
write.csv(me_ct.eur$cis$eqtls, "me_ct.eur.cis.csv", quote = F, row.names = F)
rm(me_ct.eur)

meq.eur = readRDS("meq.eur.Rds")
write.csv(meq.eur$all$eqtls, "meq.eur.csv", quote = F, row.names = F)
rm(meq.eur)

#All
me_ct.all = readRDS("me_ct.all.Rds")
write.csv(me_ct.all$trans$eqtls, "me_ct.all.tra.csv", quote = F, row.names = F)
write.csv(me_ct.all$cis$eqtls, "me_ct.all.cis.csv", quote = F, row.names = F)
rm(me_ct.all)

meq.all = readRDS("meq.all.Rds")
write.csv(meq.all$all$eqtls, "meq.all.csv", quote = F, row.names = F)
rm(meq.all)


```

### Q-Q plots

```{r}
par(cex = 0.7)
print("All")
plot(meq.all)
plot(me_ct.all)
pdf("all.p0.01.pdf")
plot(meq.all)
plot(me_ct.all)
dev.off()
pdf("all.xlim.pdf")
myPlotQQ(meq.all, xlim = c(1, 75))
dev.off()



print("EUR")
plot(meq.eur)
plot(me_ct.eur)
pdf("eur.p0.01.pdf")
plot(meq.eur)
plot(me_ct.eur)
dev.off()

print("YRI")
plot(meq.yri)
plot(me_ct.yri)
pdf("yri.p0.01.pdf")
plot(meq.yri)
plot(me_ct.yri)
dev.off()
```

## Figure 2. Table

#### Number of Identified Transposible Element (TE) eQTLs


```{r}
p.cut = 1e-2
fdr.cut = 0.05
meqs = list(EUR = meq.eur$all$eqtls, YRI = meq.yri$all$eqtls, All = meq.all$all$eqtls)
meqs.filtered = lapply(meqs, filterEQTL, p.cut = p.cut, fdr.cut = fdr.cut)
#res1 = sapply(meqs.filtered, function(x) y = table(x$snp.type))
res1 = sapply(meqs.filtered, nrow)

me_ct.cis = list(EUR = me_ct.eur$cis$eqtls, YRI = me_ct.yri$cis$eqtls, All = me_ct.all$cis$eqtls)
me_ct.cis.filtered = lapply(me_ct.cis, filterEQTL, p.cut = p.cut, fdr.cut = fdr.cut)
#res3 = sapply(me_ct.cis.filtered, function(x) y = table(x$snp.type))
res2 = sapply(me_ct.cis.filtered, nrow)

me_ct.tra = list(EUR = me_ct.eur$trans$eqtls, YRI = me_ct.yri$trans$eqtls, All = me_ct.all$trans$eqtls)
me_ct.tra.filtered = lapply(me_ct.tra, filterEQTL, p.cut = p.cut, fdr.cut = fdr.cut)
#res2 = sapply(me_ct.tra.filtered, function(x) y = table(x$snp.type))
res3 = sapply(me_ct.tra.filtered, nrow)



df = rbind(All = res1, Cis = res2, Trans = res3)

kable(df, format = "pandoc", caption = "Number of Identified Transposible Element (TE) eQTLss", padding = 2)

```


```{r}
# All
meq.all.mdf <- merge(meq.all$all$eqtls, chi.test.df, by.x = "snps", by.y = "snps")
me.all.cis.mdf <- merge(me_ct.all$cis$eqtls, chi.test.df, by.x = "snps", by.y = "snps")
me.all.tra.mdf <- merge(me_ct.all$trans$eqtls, chi.test.df, by.x = "snps", by.y = "snps")

# EUR
# meq.eur.mdf <- merge(meq.eur$all$eqtls, chi.test.df, by.x = "snps", by.y = "snps")

# YRI
# meq.yri.mdf <- merge(meq.yri$all$eqtls, chi.test.df, by.x = "snps", by.y = "snps")

#output
#write.csv(meq.all.mdf, "meq.all.csv", quote = F, row.names = F)
#write.csv(meq.eur.mdf, "meq.eur.csv", quote = F, row.names = F)
#write.csv(meq.yri.mdf, "meq.yri.csv", quote = F, row.names = F)

```

### Filter Genes of Interst



```{r}
# Sort by p-value, then by chi-square FDR adjusted q-value
meq.all.mdf = meq.all.mdf[order(meq.all.mdf$pvalue, meq.all.mdf$chisq.p), ]
me.all.cis = me.all.cis.mdf[order(me.all.cis.mdf$pvalue, me.all.cis.mdf$chisq.p), ]
me.all.tra = me.all.tra.mdf[order(me.all.tra.mdf$pvalue, me.all.tra.mdf$chisq.p), ]

# Map gene symbols
id.map = read.csv("./input/ensemblGeneToGeneName-20150206.csv")
id.map = id.map[, c("name", "gene")]
id.map.u = id.map[!duplicated(id.map$gene), ]

meq.all.mdf = merge(meq.all.mdf, id.map.u, by.x = "gene", by.y = "gene")
me.all.cis = merge(me.all.cis, id.map.u, by.x = "gene", by.y = "gene")
me.all.tra = merge(me.all.tra, id.map.u, by.x = "gene", by.y = "gene")

# Output unique genes 
#write.csv(me.all.cis, "cis-eQTLs.csv", quote = F, row.names = F)
#write.csv(me.all.tra[!duplicated(me.all.tra$gene), ], "trans-eQTLs-genelist.csv", quote = F, row.names = F)
```

### Transcription Factor as targets of cis-TE-eQTL

#### Discover cis-eQTL-TFs via Enrichmen Analysis of trans-eQTL-target

Not Working(no significant enrichment)

#### Search for cis-eQTL-TFs trans-eQTL-target based on MSigDB 

1. Look for annotated TF genes from AnimalTFDB



2. Search for TF target from MSigDB

transcription factor targets, gene symbols gene set from http://www.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/4.0/c3.tft.v4.0.symbols.gmt

```{r eval=F}
# Annotated TF and TF-target list
# tf <- read.csv("./input/AnimalTFDB-Homosapiens-20150209.csv", header = T)
# tb <- read.csv("./input/c3.tft.v4.0.symbols.gmt.csv", header = F, stringsAsFactors = F)
# tfsets <- strsplit(tb$V3, "\t")
# 
# # Cleanup TF names
# tfsets.name = tb$V1
# tfsets.name.new = gsub("^.*\\$", "", tfsets.name)
# tfsets.name.new = gsub("_.*$", "", tfsets.name.new)
# names(tfsets) = tfsets.name.new
# 
# # Find all TF in cis-eQTL-gene pairs
# me.all.cis.tf <- me.all.cis[which(me.all.cis$gene %in% tf$Ensembl.ID), ]
# me.all.cis.tf = me.all.cis.tf[order(me.all.cis.tf$chisq.q), ]
# #write.csv(me.all.cis.tf, "cis-eQTL-TF.csv", row.names = F )
# top.n = nrow(me.all.cis.tf)
# lapply(1:top.n, function(x) ploteqtl(me.all.cis.tf[x, ], by = "population"))
# 
# # Find targets of cis-eQTL-TFs
# hits <- tfsets.name.new[which(tfsets.name.new %in% me.all.cis.tf$name)]
# hits.target = tfsets[hits]
# 
# # Find trans-eQTL-targets
# tra.targets = me.all.tra[which(me.all.tra$name %in% hits.target), ]
```

As suggested by Dr Marino-Ramirez, the MSigDB is outdated and therefore, we are moving to finding the TF-gene-target pairs by doing the intersection of TFBS and upstream 2kb of gene TSS.

```{r}
library(reshape)
# Read gene name info(ID/name conversion)
# gene.info <- read.delim("./input/Homo_sapiens.gene_info.20150217", skip = 1, header = F, sep = "\t")
# gene.info = gene.info[, c(3, 5)]
# names(gene.info) = c("symbol", "synonyms")
gene.info <- read.delim("./input/BKL_export.txt", sep = "\t", header = T, stringsAsFactors = F)
colnames(gene.info) = gsub(" ", "", colnames(gene.info))
gene.info = gene.info[gene.info$Species.Taxon == "Homo sapiens", ]
gene.info = gene.info[, c(1, 5)]

## Extract only the subset of TF
# tf <- read.csv("./input/AnimalTFDB-Homosapiens-20150209.csv", header = T)
# gene.info = merge(tf, gene.info, by.x = "Symbol", by.y = "symbol")
# gene.info.syn = strsplit(gene.info$synonyms, "\\|")
# names(gene.info.syn) = gene.info$Symbol
# mdf = melt(gene.info.syn)
# names(mdf) = c("synonym", "symbol")
# gene.info = merge(mdf, gene.info, by.x = "symbol", by.y = "Symbol")
# gene.info = gene.info[, -6]

# Read Gene-TFBS pairs TRANSFAC data
transfac <- read.delim("./input/transfac_sites-hg19.gff", header = F, sep = "\t", stringsAsFactors = F)
transfac <- transfac[2:nrow(transfac), c(1, 4, 5, 9)]
names(transfac) <- c("tfbsChr", "tfbsStart", "tfbsEnd", "annotation")


# Read Gene-TFBS pairs ChIP-seq data
## Run only for the first time
##chipseq <- read.delim("./input/chip-hg19.gff", header = F, sep = "\t", stringsAsFactors = F)
##chipseq <- chipseq[2:nrow(chipseq), c(1, 4, 5, 9)]
##names(chipseq) <- c("tfbsChr", "tfbsStart", "tfbsEnd", "annotation")
##saveRDS(chipseq, "./input/chipseq.RDS")
chipseq = readRDS("./input/chipseq.RDS")

# Extract gene target and TF names
res = extractTFTG(transfac$annotation)
transfac.df = cbind(transfac[, -4], res)
transfac.pair = paste0(res$tf, ";", res$target)
transfac.pair = unique(transfac.pair)
transfac.pair = data.frame(do.call(rbind, strsplit(transfac.pair, ";")))
names(transfac.pair) = c("tf", "target")

transfac.pair = transfac.pair[grep("-miR|-let", transfac.pair$tf, perl = T, invert = T), ] #Remove miRNA entries
transfac.pair = transfac.pair[grep("_", transfac.pair$tf, invert = T), ]

# Merge TRANSFAC TF-gene pairs with gene name info to obtain HGNC symbol for TFs
mdf1 = merge(transfac.pair, gene.info, by.x = "tf", by.y = "synonym")
mdf2 = merge(transfac.pair, gene.info, by.x = "tf", by.y = "symbol")

mdf2 = mdf2[, grep("synonym", colnames(mdf2), invert = T)]
mdf1 = data.frame(tf = mdf1$symbol, mdf1[, c("target", "Ensembl.ID", "Gene.ID", "Family")])

tf.gene.df = rbind(mdf1, mdf2)



# Looking for TRANSFAC TF in cis eQTL genes while looking for targets of these TF in trans eQTL genes

cis.tf = merge(tf.gene.df, me.all.cis, by.x = "Ensembl.ID", by.y = "gene")
trans.target = merge(cis.tf, me.all.tra, by.x = "target", by.y = "name")
trans.target = trans.target[which(trans.target$snps.x == trans.target$snps.y), ]

tf.target.pair = paste0(trans.target$tf, ",", trans.target$target)
tf.target.pair = unique(tf.target.pair)
tf.target.pair = c("tf,target", tf.target.pair)

write.csv(trans.target, "cis-tf-trans-target.csv", quote = F, row.names = F)
write.csv(tf.target.pair, "cis-tf-trans-target.unique.csv", quote = F, row.names = F)
```

## Results

### All Population
- [All eQTL][11]
- [Trans eQTL][12]
- [Cis eQTL][13]
- [Q-Q plot of all eQTLs][14]
- [Q-Q plot of all eQTLs with p-value < 0.01][15]
- [Q-Q plot of cis and trans eQTLs][16]

### EUR Population
- [All eQTL][21]
- [Trans eQTL][22]
- [Cis eQTL][23]
- [Q-Q plot of all eQTLs][24]
- [Q-Q plot of all eQTLs with p-value < 0.01][25]
- [Q-Q plot of cis and trans eQTLs][26]

### YRI Population
- [All eQTL][31]
- [Trans eQTL][32]
- [Cis eQTL][33]
- [Q-Q plot of all eQTLs][34]
- [Q-Q plot of all eQTLs with p-value < 0.01][35]
- [Q-Q plot of cis and trans eQTLs][36]


```{r misc, eval = F}
## QQ plot for all p value in (0, 1)
library(gap)
tmp = read.table("meq.all.p-values.txt", header = F, stringsAsFactors = F)
observed = tmp[, 1]

qqunif(observed, pch = 16, axes = F)
axis(1, pos=0)
axis(2, pos=0)

```


[11]:./TEeQTL-paper/meq.all.csv
[12]:./TEeQTL-paper/me_ct.all.tra.csv
[13]:./TEeQTL-paper/me_ct.all.cis.csv
[14]:./TEeQTL-paper/meq.all.pdf
[15]:./TEeQTL-paper/all.p0.01.pdf
[16]:./TEeQTL-paper/me_ct.all.pdf


[21]:./TEeQTL-paper/meq.eur.csv
[22]:./TEeQTL-paper/me_ct.eur.tra.csv
[23]:./TEeQTL-paper/me_ct.eur.cis.csv
[24]:./TEeQTL-paper/meq.eur.pdf
[25]:./TEeQTL-paper/eur.p0.01.pdf
[26]:./TEeQTL-paper/me_ct.eur.pdf


[31]:./TEeQTL-paper/meq.yri.csv
[32]:./TEeQTL-paper/me_ct.yri.tra.csv
[33]:./TEeQTL-paper/me_ct.yri.cis.csv
[34]:./TEeQTL-paper/meq.yri.pdf
[35]:./TEeQTL-paper/yri.p0.01.pdf
[36]:./TEeQTL-paper/me_ct.yri.pdf
