```{r echo=FALSE}
library(knitr)
opts_chunk$set(echo=FALSE, message = FALSE)
```

---
title: "TE eQTL"
author: "Lu Wang"
date: "Tuesday, December 09, 2014"
output: html_document
---

## TE-Gene Association on All Pairs

```{r}
source("function.R")
set.seed(12345)
```

```{r cache = TRUE}
SNP_file_name = "./input/snps.txt"
snps_location_file_name = "./input/snploc.txt"
expression_file_name = "./input/geneExp.txt"
gene_location_file_name = "./input/geneloc.txt"
covariates_file_name = "./input/covariates.txt"

snps <- read.table(SNP_file_name, header = T, stringsAsFactors = F)
snploc <- read.table(snps_location_file_name, header = T, stringsAsFactors = F)
geneExp <- read.table(expression_file_name, header = T, stringsAsFactors = F)
geneloc <- read.table(gene_location_file_name, header = T, stringsAsFactors = F)
covar <- read.table(covariates_file_name, header = T, stringsAsFactors = F)

#normGeneExp <- read.table(normalized_expression_file_name, header = T, stringsAsFactors = F)
```



## Distribution of Allele Frequency

```{r cache = TRUE}

msnps <- as.matrix(snps[, grep("id", names(snps), invert = T)])
rownames(msnps) <- as.character(snps[, grep("id", names(snps))])

res <- apply(msnps, 1, table)
res1 <- sapply(res, getAF)
```

```{r}

hist(res1, breaks = 100, col = "steelblue", border = NA, main = "", xlab = "Allele Frequency")
af.df <- data.frame(names(res1), res1)
names(af.df) = c("id", "af")
af.df$id = as.character(af.df$id)
rm(res1)
```








```{r}
library(dplyr)
af.cut = 0.05
keep = which(af.df$af > af.cut)
res = snps[keep, ]
res1 = snploc[keep, ]
write.table(res, "./input/temp.snps", quote = F, row.names = F, col.names = F, sep = "\t")
write.table(res1, "./input/temp.snploc", quote = F, row.names = F, col.names = F, sep = "\t")
```

## TE-Gene Association after Applying AF cutoff = `r af.cut`

```{r cache = FALSE, results='hide'}
SNP_file_name = "./input/temp.snps"
snps_location_file_name = "./input/temp.snploc"

meh <- runME( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, useModel = modelLINEAR, verbose = F)
```

```{r}
par(cex = 0.7)
plot(meh)
```

```{r cache = FALSE, results='hide'}
meq <- runME( SNP_file_name, snps_location_file_name, expression_file_name, gene_location_file_name, covariates_file_name, useModel = modelLINEAR, "qqplot", verbose = F)
```

```{r}
par(cex = 0.7)
plot(meq)
```


```{r}
alleqtl = meh$all$eqtls

top.n = 50

lapply(1:top.n, function(x) ploteqtl(alleqtl[x, ], normalized = T, by = "population"))
```





# Breakdown by Population

```{r}
# Covariates
pop.names = c("GBR", "FIN", "CEU", "YRI", "TSI")
rownames(covar) = covar$id
df <- as.data.frame(t(covar[, grep("id", colnames(covar), invert = T)]))
barplot(table(factor(df$population, levels = 1:5, labels = pop.names)), xlab = "Population", ylab = "Number of Individuals")
df$group = as.factor(df$population)
levels(df$group) = c(2, 2, 2, 1, 2)
```


```{r breakdown, results='hide'}
sdf <- split(df, f = df$group)
res <- lapply(sdf, function(x){ df = t(x)
                                df = data.frame(id = rownames(df), df)
                                if(x$group[1] == 4){
                                  df["gender", ]
                                }else{
                                  df[c("gender", "population"), ]
                                }})

mapply(write.table, res, file = paste0("./temp3/covariates", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )
pop.covar = res

# snps
pop.names = lapply(sdf, rownames)
pop.snps = lapply(pop.names, function(x) { df = snps[, c("id", x)] } )
mapply(write.table, pop.snps, file = paste0("./temp3/snps", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )



# Expression
pop.geneExp = lapply(pop.names, function(x) { df = geneExp[, c("id", x)] } )
mapply(write.table, pop.geneExp, file = paste0("./temp3/geneExp", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )

# geneloc
pop.genes = lapply(pop.geneExp, function(x) as.character(x$id))
rownames(geneloc) = geneloc$id
pop.geneloc = lapply(pop.genes, function(x) { df = geneloc[x, ] } )
mapply(write.table, pop.geneloc, file = paste0("./temp3/geneloc", 1:2, ".txt"), quote = rep(FALSE, 2), row.names = rep(FALSE, 2), sep = rep("\t", 2) )
```


## European Population (GBR, FIN, CEU, TSI)

```{r population1, results='hide'}
af.cut = 0.05

x = 1

res <-  runPop(x = x, af.cut = af.cut, useModel = modelLINEAR, verbose = F)

```


```{r}
par(cex = 0.7)
plot(res$meh)
plot(res$meq)
```

```{r}
alleqtl.eu = res$meq$all$eqtls


lapply(1:top.n, function(x) ploteqtl(alleqtl.eu[x, ], normalized = T, by = "population"))
```


## African Population (YRI)

```{r population2, cache = TRUE, results='hide', eval = FALSE}
af.cut = 0.05

x = 2

res <-  runPop(x = x, af.cut = af.cut, useModel = modelLINEAR, verbose = F)

```


```{r, eval = FALSE}
par(cex = 0.7)
plot(res$meh)
plot(res$meq)
```

```{r}
alleqtl.yri = res$meq$all$eqtls

lapply(1:top.n, function(x) ploteqtl(alleqtl.yri[x, ], normalized = T, by = "population"))
```

